File: 1717/solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public int maximumGain(String s, int x, int y) {
        int res = 0;
        List<Character> stack = new ArrayList<>();
        String goodVal = x > y ? "ab" : "ba";
        int small = Math.min(x, y);
        int big = Math.max(x, y);

        for (char c : s.toCharArray()) {
            if (c > 'b') {
                res += calculateRemainingStack(stack, small);
                stack.clear();
            } else if (!stack.isEmpty() && c == goodVal.charAt(1) && stack.getLast() == goodVal.charAt(0)) {
                stack.removeLast();
                res += big;
            } else {
                stack.add(c);
            }
        }

        if (!stack.isEmpty()) {
            res += calculateRemainingStack(stack, small);
        }
        return res;
    }

    private int calculateRemainingStack(List<Character> stack, int val) {
        int ays = 0;
        int bees = 0;
        for (char c : stack) {
            if (c == 'a') {
                ays++;
            } else {
                bees++;
            }
        }
        return Math.min(ays, bees) * val;
    }
}

// time O(n)
// space O(n)

File: 1717/solution.ts
function maximumGain(s: string, x: number, y: number): number {
  let res = 0;
  let stack: Array<string> = [];
  const goodVal = x > y ? "ab" : "ba";
  const small = Math.min(x, y);
  const big = Math.max(x, y);

  for (const c of s) {
    if (c > "b") {
      res += calculateRemainingStack(stack, small);
      stack = [];
    } else if (
      stack.length &&
      c == goodVal.charAt(1) &&
      stack[stack.length - 1] == goodVal.charAt(0)
    ) {
      stack.pop();
      res += big;
    } else {
      stack.push(c);
    }
  }

  if (stack.length) {
    res += calculateRemainingStack(stack, small);
  }
  return res;
}

const calculateRemainingStack = (stack: Array<string>, val: number): number => {
  let ays = 0;
  let bees = 0;
  for (const c of stack) {
    if (c == "a") {
      ays++;
    } else {
      bees++;
    }
  }
  return Math.min(ays, bees) * val;
};
